Updating Limiters for Finite Volume Formulation

tanoret opened the pull request:
Closes #28891

Commit:
changing second order upwind and limitng formulation to work with incompressible flow Refs #28891

tanoret commented:
@freiler plase also take a look if you can

tanoret commented:
This are the current results on October 18, 2024

Upwind

<img width="604" alt="image" src="https://github.com/user-attachments/assets/d4627c3d-a6f0-4132-9525-d93813b85967">

Second-Order Upwind:

<img width="604" alt="image" src="https://github.com/user-attachments/assets/58d63659-76bc-462e-be27-adb30511670c">

Min-Mod:

<img width="604" alt="image" src="https://github.com/user-attachments/assets/56ec81d5-4332-44b3-b002-581177c5e12d">

VanLeer:

<img width="604" alt="image" src="https://github.com/user-attachments/assets/34ee30ba-271d-4116-a0f5-db68f4ec6027">

QUICK:

<img width="604" alt="image" src="https://github.com/user-attachments/assets/3076d27a-b881-4dec-8206-5ed3b0f0a216">

Venkatakrishnan:

<img width="604" alt="image" src="https://github.com/user-attachments/assets/c56cd6e0-b03c-4e1b-8d32-b4b3da6aad6b">


Notes:

- Flux limiters are lagged in time step
- Deterioration in convergence is only observed for Venkatakrishnan (all the rest converge as upwind)
- QUICK does not show improvements with rest to upwind

tanoret commented:
To-do:
- [ ] Modify the implicit vs explicit update in the limiters and select the best default option for each
- [ ] Correct QUICK issues
- [ ] Improve the documentation for all functions in limiters and MathFVUtils
- [ ] Write updated documentation on limiters

GiudGiud commented:
you'll want to get rid of the submodule update in the wasp folder

tanoret committed:
correcting limting action to face grdient limiter Refs #28891

tanoret committed:
amending limiter documentation Refs #28891

lindsayad commented on framework/include/base/MooseFunctorArguments.h:
Can we remove the reorganization portion of this diff?

tanoret replied:
It is hard to get rid of this diff since I has to move `FaceArg` down the header as it is now taking a `Moose::StateArg` that defines the state in which the limiters are computed. Is there any way that I can force git not to take a diff?

lindsayad reviewed (commented):
Should I wait to review this until you've removed the WIP label? In which case should we convert this to draft?

tanoret committed:
adding limiter tests Refs #28891

tanoret commented:
This are the latest results. Images are in the docs. Limiters are working as expected. Tests include:

- advection in a cartesian mesh in two directions (bottom-left to top-right and inverted)
- advection in a triangular mesh
- lid-driven cavity with Newton
- lid-driven cavity with SIMPLE

![image](https://github.com/user-attachments/assets/e16ebae8-5b67-4e09-b05c-ab7d33e75f95)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2510505/), step Clang format on 05c8cfe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b3924ece4ed556a7ca9811a6e046eb5c4503df72`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 19f9bb9a092270d4a317db0c6d21a7a495af1e27
+Subproject commit 98c8751a0d2907d235d8e1e9ae6c54a867ab9380
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2510505/), step Size check on 05c8cfe wanted to post the following:

Warning: This PR changes repo size by 5.26 MiB.

lindsayad commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,280 @@ class Limiter
   virtual InterpMethod interpMethod() const = 0;
+
+  /**
+   * @brief Functor for applying tsimplified slope limiting.
```
```suggestion
   * @brief Functor for applying simplified slope limiting.
```

tanoret replied:
It is hard to get rid of this diff since I has to move `FaceArg` down the header as it is now taking a `Moose::StateArg` that defines the state in which the limiters are computed. Is there any way that I can force git not to take a diff?

tanoret committed:
style changes required by code pre-check and removing wasp Refs #28891

lindsayad commented:
You can mention me once you're done pushing and I will review. Going to unsubscribe for now

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2513487/), step Size check on 49dd7f5 wanted to post the following:

Warning: This PR changes repo size by 5.26 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2516181/), step Size check on f5e60f7 wanted to post the following:

Warning: This PR changes repo size by 5.26 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2516273/), step Size check on 149d840 wanted to post the following:

Warning: This PR changes repo size by 5.26 MiB.

GiudGiud commented:
We'll want to reduce the size of the tests. 25x25 for a full grid of tests is bloating the repo. You ll see it fail the same way with 5x5 but at (theoretically) 1/25th of the memory cost

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2516268/), step Docs: sync website on 149d840 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28892/site)

This comment will be updated on new commits.

GiudGiud commented on framework/include/base/MooseFunctorArguments.h:
```diff
@@ -75,6 +75,94 @@ struct ElemPointArg
+ * nonlinear iteration, etc.
+ */
+
+struct StateArg
```
why move this up ?

tanoret committed:
chaging interpolation method for vectors in MathFVUtils and regolding vanLeer vector limiter (ref difs ~1e-5) Refs #28891

GiudGiud reviewed (commented):
early feedback, not quite done but might be delayed tomorrow

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2518838/), step Size check on 076ba98 wanted to post the following:

Warning: This PR changes repo size by 5.28 MiB.

tanoret commented:
For the `fvkernels/mms/advective-outflow.QUICKLimiter`, which is supposed to test QUICK convergence with order $2 \pm 0.05$, we are converging with order 2.3. Does someone now rapidly what could be going on with this test or should I dig deeper into this?

![image](https://github.com/user-attachments/assets/5d0e9488-ff68-4f19-ae13-1dbb2fb0556f)

tanoret committed:
changing fvkernels/mms/advection-outflow test logic to ensure that convergence is at leaste the required order of convergence Refs #28891

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519576/), step Size check on 15559c5 wanted to post the following:

Warning: This PR changes repo size by 5.28 MiB.

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,280 @@ class Limiter
   }
 
+  /**
+   * @brief Computes the flux limiting ratio using gradients.
```
Yes. I added a clarification

tanoret replied:
Well, partially, the previous method had linear gradients only (i.e., not GG or LS). So, the limiting was not really working well for incompressible and weakly-compressible flow

GiudGiud replied:
well you say it works well right below in the text??

tanoret replied:
It works now with the new mods

tanoret replied:
It did not work in the previous formulation

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,280 @@ class Limiter
+   * @param dCD A constant direction vector representing the direction of the cell face.
+   * @return The computed flux limiting ratio.
+   *
+   * This function performs the following steps:
```
Yes. Thanks. Removed

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,280 @@ class Limiter
+   * @note The small epsilon value `1e-10` is added to the delta max and delta min values to
+   *       avoid division by zero and numerical instability.
+   *
+   * @example
+   * @code
+   * Real phi_upwind = 2.0;
+   * VectorValue<Real> grad_upwind(0.1, 0.2, 0.3);
+   * Real max_value = 5.0;
+   * Real min_value = 1.0;
+   * FaceInfo * fi = ... // Assume this is properly initialized
+   * bool is_upwind = true;
+   * Real ratio = rf_minmax(phi_upwind, &grad_upwind, max_value, min_value, fi, is_upwind);
+   * @endcode
```
I was hoping we could mantain them too. Adding a utilization example is generally industry standard for clean code. Since you guys will be the main maintainers of this I am good with whatever you decide, but let's give it one more deep thought please

tanoret replied:
Ok. So we won't document how things are used. No problem. Removed

tanoret commented on framework/include/limiters/MinModLimiter.h:
```diff
@@ -17,27 +17,95 @@ namespace Moose
     return 0 * r_f + std::max(T(0), std::min(T(1), r_f));
   }
+
+  /**
+   * @brief Indicates whether the Min-Mod limiter is constant.
+   *
+   * This method always returns `false` as the Min-Mod limiter is not a constant limiter.
+   *
+   * @return `false` indicating the Min-Mod limiter is not constant.
+   */
```
Isn't it easier for a future developer to understand what is going on if every function is properly documented in the context in which it is use?

GiudGiud replied:
developer should know to look at the base class header for the docstrings, if they are not looking at the doxygen already

tanoret replied:
As per the comment above, removing extra context is a waste of time (and also makes usability and documentation for new users poorer). I will remove it now, but I really don't agree with this

tanoret commented on framework/include/limiters/MinModLimiter.h:
```diff
@@ -17,27 +17,95 @@ namespace Moose
 class MinModLimiter : public Limiter<T>
 {
 public:
+  /**
```
Are you sure? This was not showing the previous doxygen for the MinMod class (https://mooseframework.inl.gov/docs/doxygen/moose/classMoose_1_1FV_1_1MinModLimiter.html)

I like to document properly what every class is doing, but, again, since you will ba mantaining this code on more regular basis, I am open to delete. Please just give it one more thought

GiudGiud replied:
it shows here?
https://mooseframework.inl.gov/docs/doxygen/moose/classMoose_1_1FV_1_1MinModLimiter.html#afbe7bc5b8d9caee6b0355721d44dc7f1

it's really insufficient right now, but that's because Limiter.h only has this one sentence for Limit()

loganharbour replied:
In my opinion asking someone to remove extra context is a waste of time

tanoret replied:
Yes. I agree with this last comment. I am keeping this one

tanoret commented on modules/navier_stokes/src/fvkernels/INSFVEnergyAdvection.C:
```diff
@@ -36,9 +36,14 @@ ADReal
 INSFVEnergyAdvection::computeQpResidual()
 {
   const auto v = velocity();
+  const auto & limiter_time = _subproblem.isTransient()
+                                  ? Moose::StateArg(1, Moose::SolutionIterationType::Time)
+                                  : Moose::StateArg(1, Moose::SolutionIterationType::Nonlinear);
```
Because I don't have an old time step and if I limit implicitly the whole limiting theory breaks and the condition number of the Jacobian is really bad. Note that this works for both Newton steay and nonlinear SIMPLE. In brief, we wants a limiter that converges as good as upwind while still limiting

tanoret commented on modules/navier_stokes/test/tests/finite_volume/limiters/dispersion-test/tests:
```diff
@@ -0,0 +1,104 @@
+    []
+  []
+  [top_right_limited_scalar_advection]
+    requirement = 'The system shall be able to perform a variety of limiting schemes when solving scalar transport equations in cartesian meshes with top-right advection. These schemes include'
```
Yes. The gradient limiting is directional. I just want to make sure that no direction is ever messed up

GiudGiud replied:
The gradient limiting is directional as in it depends on which direction is upwind. high/low X, Y, Z are not coded anywhere?

tanoret replied:
No. It also has a directional component when computing the gradient at the center of the cell. If someone messess this up it needs to show in these tests

GiudGiud replied:
@grmnptr do you understand why that would be?
I m at a loss on this.

tanoret replied:
It should be in the documentation. Please let me know what you are not following and I am happy to explain

GiudGiud replied:
how is top right any different from bottom left in the code?
we dont hard code any direction as different from any other?

GiudGiud replied:
ok I see now what you are thinking of. 
Unfortunately we cant be targeting each potential mistake in modifications using regression tests. For example, what if someone switches the Y and X components? Both of these 45 degree tests would not catch it, yet we wont add 5 more tests make sure. 

Testing the limiter for such behavior should be done in a unit test if you care to prevent it.

tanoret replied:
Thanks. All limiters don't work the same and depend on the gradient somehow. So, I will still need to test them in both directions. I can remove the upwind test for advection from the top-right if you like. It is the only redundant one that I can think of

tanoret commented on modules/navier_stokes/test/tests/finite_volume/limiters/lid-driven/tests:
```diff
@@ -1,8 +1,15 @@
+      input = 'test.i'
+      exodiff = upwind.e
+      cli_args = 'GlobalParams/advected_interp_method=upwind Outputs/file_base=upwind'
+      detail = 'first-order upwind'
```
You may see it that way. You just take the limiter equal to zero ($\beta = 0$) to recover upwind. I will remove if you think it is not appropaite

tanoret replied:
Keeping the test for completness

tanoret committed:
Apply suggestions from giudgiud code review 1 Refs #28891

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

regolding tests to latest mods in limiters and avoid testing unnecessary recovers Refs #28891

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519807/), step Clang format on 1bc9eb5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fe524dec2ce86e8850eb17d974b81e661d28b1d2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519807/), step Size check on 1bc9eb5 wanted to post the following:

Warning: This PR changes repo size by 5.28 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519807/), step Clang format on 1bc9eb5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fe524dec2ce86e8850eb17d974b81e661d28b1d2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519807/), step Size check on 1bc9eb5 wanted to post the following:

Warning: This PR changes repo size by 5.28 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519858/), step Clang format on 83e44c5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fe524dec2ce86e8850eb17d974b81e661d28b1d2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2519858/), step Size check on 83e44c5 wanted to post the following:

Warning: This PR changes repo size by 5.28 MiB.

tanoret committed:
extending supported two-term-boundary expansion methods in INSFAdvectionKernel  Refs #28891

tanoret committed:
completing argument definition for passing compile checks Refs #28891

tanoret committed:
completing argument definition for passing compile checks and modifying test tolerance Refs #28891

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2524302/), step Clang format on d72d4da wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b8db753c73821a52d1624333d28c3c8bf9b2a008`

GiudGiud commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,280 @@ class Limiter
+   * @brief Functor for applying simplified slope limiting.
+   *
+   * This function applies the limiter by invoking the `limit` method with the provided parameters.
+   * It acts as a functor, enabling objects of the `VanLeerLimiter` class to be used as if they were
```
this is the doco for Limiter not VanLeer

GiudGiud commented on framework/include/limiters/VenkatakrishnanLimiter.h:
```diff
@@ -0,0 +1,140 @@
+ *
+ * 4. Venkatakrishnan limiter formula:
+ * \f[
+ * \beta(r_f) = \frac{2 r_f + 1.0}{r_f (2 r_f + 1.0) + 1.0}
```
do you have a reference for this formula? 
I m seeing a coefficient K in the Ventakrishnan limiter on random references

tanoret replied:
Yes. It is the one that I put in markdown file. Adding here for completness. As per Sewby's diagram, the K factor different than one deteriorates boundness

GiudGiud commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,442 @@ interpolate(const Limiter & limiter,
+computeMinMaxValue(const FunctorBase<T> & functor, const FaceArg & face, const StateArg & time)
+{
+  // Initialize max_value to 0 and min_value to a large value (1e10)
+  T max_value(0), min_value(1e10);
```
advected quantity could go negative if it's momentum ? so a starting max_value of 0 isnt reliable

tanoret replied:
Chaneged by Real bounds. This is a good comment

GiudGiud commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,442 @@ interpolate(const Limiter & limiter,
+  }
+
+  // Iterate over the neighbors of the neighbor associated with the face
+  for (auto neighbor : (*face.fi).neighbor().neighbor_ptr_range())
```
```suggestion
  for (const auto neighbor : (*face.fi).neighbor().neighbor_ptr_range())
```

tanoret replied:
Corrected

GiudGiud commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,442 @@ interpolate(const Limiter & limiter,
+  }
+
+  // Iterate over the neighbors of the neighbor
+  for (auto neighbor : (*face.fi).neighbor().neighbor_ptr_range())
```
```suggestion
  for (const auto neighbor : (*face.fi).neighbor().neighbor_ptr_range())
```

tanoret replied:
Corrected

GiudGiud commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,442 @@ interpolate(const Limiter & limiter,
+  T max_value(0), min_value(1e10);
+
+  // Iterate over the direct neighbors of the element associated with the face
+  for (auto neighbor : (*face.fi).elem().neighbor_ptr_range())
```
```suggestion
  for (const auto neighbor : (*face.fi).elem().neighbor_ptr_range())
```

tanoret replied:
Corrected

GiudGiud commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,442 @@ interpolate(const Limiter & limiter,
+  T max_value(0), min_value(1e10);
+
+  // Iterate over the direct neighbors of the element associated with the face
+  for (auto neighbor : (*face.fi).elem().neighbor_ptr_range())
```
```suggestion
  for (const auto neighbor : (*face.fi).elem().neighbor_ptr_range())
```

tanoret replied:
Corrected

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/limiters/dispersion-test/gold/bottom_left_min_mod.e:
unfortunately you also need to delete the 230kb file from the commit history. Otherwise it's added to the repository the exact same way than if you had kept it

tanoret commented on framework/include/limiters/QUICKLimiter.h:
```diff
@@ -17,26 +17,117 @@ namespace Moose
+  /**
+   * @brief Indicates whether the QUICK limiter is constant.
+   *
+   * This method always returns `false` as the QUICK limiter is not a constant limiter.
```
This is how functions are documented in production standard. Base text was written by user, formating and spelling check details were improved by Continue (INL-approved tool for VS Code). I really don't know what to reply to this comment...

GiudGiud replied:
Thanks for confirming. 
There's nothing special to reply. We should tune the behavior of Continue, this bloating of docstrings is not what we want.

tanoret commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -653,6 +653,9 @@ class MooseVariableFV : public MooseVariableField<OutputType>
                                                   const Elem * elem_side_to_extrapolate_from,
                                                   const StateArg & state) const;
 
+  /// Function to get the two term boundary expansion for the variable
```
Corrected

tanoret commented on test/tests/fvkernels/mms/advective-outflow/test.py:
```diff
@@ -1,6 +1,8 @@
-from mooseutils import fuzzyAbsoluteEqual
+
+def bottom_bound(value, order, tolerance):
+    return value > (order - tolerance)
```
No. If something increases the order it breaks the theory and could be error cancelation. We should gurantee at least the order of convergence computed analytically. I am not removing this one

GiudGiud replied:
so if it breaks the theory, it s not a good idea to accept it. Which you are right now by only checking the bottom bound

tanoret replied:
It is error cancelation. We cannot really avoid that. I can increase the number of points that we use for the fitting to get closer to 2 in the order of convergence, but I don't really see a point on doing that. It is just easier to check that the order of convergence is at least the theoretical one

tanoret committed:
addressing guiallaume's code-reviewer comments Refs #28891

tanoret committed:
changing test tolerances to match serial and parallel runs Refs #28891

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2539573/), step Generate coverage on 3f39108 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>681f13</th>
      <th colspan=4>#28892 3f3910</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.11%</td>
      <td>85.07%</td>
      <td>-0.04%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106975</td>
      <td>107070</td>
      <td>+95</td>
      <td>148</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18716</td>
      <td>18787</td>
      <td>+71</td>
      <td>74</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28892/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28892/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>681f13</th>
      <th colspan=4>#28892 3f3910</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.81%</td>
      <td>88.81%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4404</td>
      <td>4404</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>555</td>
      <td>555</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28892/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28892/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>681f13</th>
      <th colspan=4>#28892 3f3910</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.64%</td>
      <td>84.65%</td>
      <td>+0.01%</td>
      <td>98.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17081</td>
      <td>17097</td>
      <td>+16</td>
      <td>49</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3100</td>
      <td>3101</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28892/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28892/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28892/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 66.67% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/contrib/wasp:
no need for this change

GiudGiud replied:
fixed already

GiudGiud commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+               const VectorValue<T> * grad_phi_upwind,
+               const VectorValue<T> * grad_phi_downwind,
+               const RealVectorValue & dCD,
+               const Real & max_value,
+               const Real & min_value,
```
these two should still be T

if we use this with a vector for phi_upwind, we wont be use the min/max Real

tanoret replied:
It does not compile with T without significant refactoring. min/max will always ultimately default to real values

tanoret replied:
I think that I am keeping these as Real. It does not really makes sense that the min/max value in the limiter is a vector...

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/limiters/dispersion-test/tests:
```diff
@@ -0,0 +1,130 @@
+      cli_args = 'GlobalParams/advected_interp_method=sou Outputs/file_base=bottom_left_sou'
+      rel_err = 1e-4
+      detail = 'second-order upwind'
+      recover = false
```
if it diffs in recover we need to know why, or at least have an issue open

tanoret replied:
There is an issue that restart tries to get the file as the other tests are running. This in mostly an issue with the test system rather than this specific test

tanoret replied:
Thanks. Corrected

GiudGiud reviewed (changes requested):
let's squash the regold commit with the commit that introduces those files. If we dont do that, we would check in both versions of the file in the repo

GiudGiud commented:
Numerical diffusivity seems to have improved a lot for SOU and Minmod. Quick and Ventakakrishnan seem to still be very diffusive. Do we expect this for these two? Is there something we can reference? Maybe this classic case has been studied with another code and those limiters in the litterature?

tanoret commented:
> Numerical diffusivity seems to have improved a lot for SOU and Minmod. Quick and Ventakakrishnan seem to still be very diffusive. Do we expect this for these two? Is there something we can reference? Maybe this classic case has been studied with another code and those limiters in the litterature?

This assessment is not correct. Taking a screenshot of the comment above showing how numerical diffusivity behaves

<img width="639" alt="image" src="https://github.com/user-attachments/assets/bdc8bb06-3777-46bf-bbd9-ef21bed39e2c">

GiudGiud commented:
so the figures in your first post are outdated?

tanoret commented:
No. They are not. What makes you think that? The tests are in a very corse mesh. The actual limiting tests are done for a finer mesh

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/limiters/dispersion-test/bottom_left_tri_mesh_advection.i:
```diff
@@ -0,0 +1,116 @@
+[GlobalParams]
```
tri mesh is still a good idea to test

tanoret replied:
Thanks

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/limiters/dispersion-test/bottom_left_advection.i:
```diff
@@ -0,0 +1,101 @@
+  [out]
+    type = Exodus
+    execute_on = 'final'
+  []
```
```suggestion
    hide = 'vel_x vel_y pressure'
  []
```
same for all the other tests, these fields are just flat

tanoret replied:
Thanks. Will change

tanoret replied:
Changed

tanoret commented:
> because on your figure at the top of this thread, these two (Venta, QUICK) have results on that 2D case that look very much like 1st order upwind, while SOU, Vanleer and MinMod are very distinctly less numerically diffusive. Do we have MMS tests that check the order at this point?

Yes. It is in the tests (test/tests/fvkernels/mms). There is no need to test every single limiter though. Order of convergence is not related to dipersion

grmnptr reviewed (changes requested):
Nice addition thanks! The main observations/comments:
- The dispersion tests look great!
- We fell on the other side of the documentation spectrum. It is good to have enough documentation but adding a summary and usage is a bit too much considering the definitions are very nicely documented and the variables and the functions are very nicely named. Doxygen will show examples and documentation along the inheritance lines. If there is a specific need to explain what is happening besides the source comments, I'd do it in a design markdown file.
- This will likely go through a rewrite when we (hopefully) transition towards an object-oriented fv interpolation paradigm. So I think we should maximize our chances (using tight tolerances) of catching issues when that happens.

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   * @return The computed flux limiting ratio.
+   *
+   * This function performs the following steps:
+   * 1. Computes the dot product of the upwind gradient vector with the direction vector `dCD`.
+   * 2. Computes the dot product of the downwind gradient vector with the direction vector `dCD`.
+   * 3. Calculates the ratio of the upwind gradient dot product to the downwind gradient dot
+   * product, adding a small epsilon value to the denominator to prevent division by zero.
+   *
+   * @note The small epsilon value `1e-10` is added to the denominator to avoid division by zero
+   *       and numerical instability.
+   *
+   * @example
+   * @code
+   * VectorValue<Real> grad_upwind(0.1, 0.2, 0.3);
+   * VectorValue<Real> grad_downwind(0.4, 0.5, 0.6);
+   * RealVectorValue dCD(1.0, 0.0, 0.0);
+   * Real ratio = rf_grad(&grad_upwind, &grad_downwind, dCD);
+   * @endcode
```
Thanks. Will remove as per our private conversation.

tanoret replied:
Removed

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   * @param fi_elem_is_upwind Boolean indicating if the face info element is upwind.
+   * @return The computed flux limiting ratio.
+   *
+   * This function performs the following steps:
+   * 1. Retrieves the face centroid from the `FaceInfo` object.
+   * 2. Determines the cell centroid based on whether the current element is upwind.
+   * 3. Computes the delta value at the face by taking the dot product of the upwind gradient
+   *    vector with the difference between the face and cell centroids.
+   * 4. Computes the delta values for the maximum and minimum allowable values, adding a small
+   *    epsilon value to prevent division by zero.
+   * 5. Calculates the flux limiting ratio based on whether the delta value at the face is
+   *    non-negative or negative, using the appropriate delta (either max or min).
+   *
+   * @note The small epsilon value `1e-10` is added to the delta max and delta min values to
+   *       avoid division by zero and numerical instability.
```
Thanks. Will remove

tanoret commented on framework/include/limiters/MinModLimiter.h:
```diff
@@ -17,27 +17,81 @@ namespace Moose
 {
 public:
+  /**
+   * @brief Computes the limited value using the Min-Mod limiter.
+   *
+   * This method overrides the pure virtual `limit` method in the base `Limiter` class.
+   * It calculates the flux limiting ratio based on the Min-Mod limiter formula.
+   *
+   * @param grad_phi_upwind Pointer to the gradient vector at the upwind location.
+   * @param grad_phi_downwind Pointer to the gradient vector at the downwind location.
+   * @param dCD A constant direction vector representing the direction of the cell face.
+   * @return The computed flux limiting ratio.
+   *
+   * This method performs the following steps:
+   * 1. Asserts that the upwind gradient pointer is not null.
+   * 2. Computes the gradient ratio coefficient \( r_f \) using the `rf_grad` method.
+   * 3. Applies the Min-Mod limiter formula to \( r_f \) to obtain the limited value.
+   * 4. Returns the computed limited value.
+   *
+   * @example
+   * @code
+   * MinModLimiter<Real> minMod;
+   * VectorValue<Real> grad_upwind(0.1, 0.2, 0.3);
+   * VectorValue<Real> grad_downwind(0.4, 0.5, 0.6);
+   * RealVectorValue dCD(1.0, 0.0, 0.0);
+   * Real result = minMod.limit(0.0, 0.0, &grad_upwind, &grad_downwind, dCD, 0.0, 0.0, nullptr,
+   * true);
+   * @endcode
+   */
```
I thought you asked me to remove this from the base class above...

tanoret replied:
Removed also from here

tanoret commented on framework/include/limiters/QUICKLimiter.h:
```diff
@@ -17,26 +17,103 @@ namespace Moose
 class QUICKLimiter : public Limiter<T>
 {
 public:
+  /**
+   * @brief Computes the limited value using the QUICK limiter.
+   *
+   * This method overrides the pure virtual `limit` method in the base `Limiter` class.
+   * It calculates the flux limiting ratio based on the QUICK limiter formula.
+   *
+   * @param phi_upwind Scalar value at the upwind location.
+   * @param phi_downwind Scalar value at the downwind location.
+   * @param grad_phi_upwind Pointer to the gradient vector at the upwind location.
+   * @param grad_phi_downwind Pointer to the gradient vector at the downwind location.
+   * @param dCD A constant direction vector representing the direction of the cell face.
+   * @return The computed flux limiting ratio.
+   *
+   * This method performs the following steps:
+   * 1. Asserts that the upwind gradient pointer is not null.
+   * 2. Computes the gradient ratio coefficient \( r_f \) using the `rf_grad` method or `rF` method.
+   * 3. Applies the QUICK limiter formula to \( r_f \) to obtain the limited value, ensuring TVD
+   * compliance.
+   * 4. Returns the computed limited value.
+   *
+   * @example
+   * @code
+   * QUICKLimiter<Real> quick;
+   * VectorValue<Real> grad_upwind(0.1, 0.2, 0.3);
+   * VectorValue<Real> grad_downwind(0.4, 0.5, 0.6);
+   * RealVectorValue dCD(1.0, 0.0, 0.0);
+   * Real result = quick.limit(0.0, 0.0, &grad_upwind, &grad_downwind, dCD, 0.0, 0.0, nullptr,
+   * true);
+   * @endcode
+   */
```
Thanks. Idem than above

tanoret replied:
Removed

tanoret commented on framework/include/limiters/SOULimiter.h:
```diff
@@ -17,25 +17,108 @@ namespace Moose
 class SOULimiter : public Limiter<T>
 {
 public:
+  /**
+   * @brief Computes the limited value using the SOU limiter.
+   *
+   * This method overrides the pure virtual `limit` method in the base `Limiter` class.
+   * It calculates the flux limiting ratio based on the SOU limiter formula.
+   *
+   * @param grad_phi_upwind Pointer to the gradient vector at the upwind location.
+   * @param grad_phi_downwind Pointer to the gradient vector at the downwind location.
+   * @param dCD A constant direction vector representing the direction of the cell face.
+   * @param max_value The maximum value in the current element.
+   * @param min_value The minimum value in the current element.
+   * @param fi Pointer to the face information structure.
+   * @param fi_elem_is_upwind Boolean flag indicating if the current element is upwind.
+   * @return The computed flux limiting ratio.
+   *
+   * This method performs the following steps:
+   * 1. Asserts that the upwind gradient pointer is not null.
+   * 2. Handles initialization conflict by determining the face centroid and the appropriate cell
+   * centroid.
+   * 3. Computes the absolute delta value at the face.
+   * 4. Computes the delta between the two elements.
+   * 5. Returns the limited value based on the computed deltas.
+   *
+   * @example
+   * @code
+   * SOULimiter<Real> sou;
+   * VectorValue<Real> grad_upwind(0.1, 0.2, 0.3);
+   * RealVectorValue dCD(1.0, 0.0, 0.0);
+   * Real max_value = 1.0;
+   * Real min_value = 0.0;
+   * FaceInfo fi;
+   * bool fi_elem_is_upwind = true;
+   * Real result = sou.limit(0.0, 0.0, &grad_upwind, nullptr, dCD, max_value, min_value, &fi,
+   * fi_elem_is_upwind);
+   * @endcode
+   */
```
Idem

tanoret replied:
Removed

tanoret commented on framework/include/limiters/VanLeerLimiter.h:
```diff
@@ -17,25 +17,81 @@ namespace Moose
 class VanLeerLimiter : public Limiter<T>
 {
 public:
+  /**
+   * @brief Computes the limited value using the Van Leer limiter.
+   *
+   * This method overrides the pure virtual `limit` method in the base `Limiter` class.
+   * It calculates the flux limiting ratio based on the Van Leer limiter formula.
+   *
+   * @param grad_phi_upwind Pointer to the gradient vector at the upwind location.
+   * @param grad_phi_downwind Pointer to the gradient vector at the downwind location.
+   * @param dCD A constant direction vector representing the direction of the cell face.
+   * @return The computed flux limiting ratio.
+   *
+   * This method performs the following steps:
+   * 1. Asserts that the upwind and downwind gradient pointers are not null.
+   * 2. Computes the gradient ratio coefficient \( r_f \) using the `rf_grad` method.
+   * 3. Applies the Van Leer limiter formula to \( r_f \) to obtain the limited value.
+   * 4. Returns the computed limited value.
+   *
+   * @example
+   * @code
+   * VanLeerLimiter<Real> vanLeer;
+   * VectorValue<Real> grad_upwind(0.1, 0.2, 0.3);
+   * VectorValue<Real> grad_downwind(0.4, 0.5, 0.6);
+   * RealVectorValue dCD(1.0, 0.0, 0.0);
+   * Real result = vanLeer.limit(0.0, 0.0, &grad_upwind, &grad_downwind, dCD, 0.0, 0.0, nullptr,
+   * true);
+   * @endcode
+   */
```
Idem

tanoret replied:
Removed

tanoret commented on framework/include/limiters/VanLeerLimiter.h:
```diff
@@ -17,25 +17,81 @@ namespace Moose
           const T & phi_downwind,
           const VectorValue<T> * grad_phi_upwind,
-          const RealVectorValue & dCD) const override final
+          const VectorValue<T> * grad_phi_downwind,
+          const RealVectorValue & dCD,
+          const Real & /* max_value */,
+          const Real & /* min_value */,
+          const FaceInfo * /* fi */,
```
Thanks. That would be better

tanoret commented on modules/navier_stokes/test/tests/finite_volume/limiters/dispersion-test/top_right_advection.i:
```diff
@@ -0,0 +1,102 @@
+[GlobalParams]
```
Thanks. Will update

tanoret replied:
Updated

tanoret commented on modules/navier_stokes/test/tests/finite_volume/limiters/lid-driven-segregated/tests:
```diff
@@ -0,0 +1,75 @@
+      input = 'lid-driven-segregated.i'
+      exodiff = 'upwind.e'
+      detail = 'and reach converged results with upwind advection scheme.'
+      rel_err = 1e-3
```
Thanks. I tried slowly increasing it until the tests passed. Could not figure it out with `abs_tol`. If codes change things 1% or 0.1% I really don't mind for practical purposes. I will give it one more spin jic

tanoret replied:
Thanks. Tightened the tolerances as much as possible for the parallel tests to still run

tanoret commented on modules/navier_stokes/test/tests/finite_volume/limiters/lid-driven/test.i:
```diff
@@ -128,13 +131,20 @@ rho=1
 
 [Executioner]
-  type = Steady
+  type = Transient
```
Because of the way in which we limit using the explicit limiting flux computed from the previous step. Steady solves won't be really getting the right limiting

tanoret replied:
Add warning in the documentation

tanoret reviewed (commented):
Thanks. Adding promised changes in the upcoming commit

tanoret committed:
Apply suggestions from code review peter Refs #28891

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   *         return ratio; // Return the computed ratio.
+   *     }
+   * };
+   * @endcode
```
Thanks removed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2548082/), step Clang format on 9fe0a7c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7377410027ba9dbf48902235a0437fdf96485e29`

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -633,27 +1090,91 @@ interpolate(const FunctorBase<VectorValue<T>> & functor,
+ * access to neighboring elements and limiter type.
+ * @param time An argument representing the state or time at which the evaluation is performed.
+ *
+ * @return T The interpolated container value at the face.
+ *
+ * This function performs the following steps:
+ * 1. Asserts that the face information is non-null.
+ * 2. Constructs a limiter based on the face limiter type.
+ * 3. Determines the upwind and downwind arguments based on the face element.
+ * 4. Evaluates the functor at the upwind and downwind locations.
+ * 5. Initializes the return container value.
+ * 6. Computes the interpolation coefficients and advected values for each component.
+ * 7. If necessary, computes the gradient and min-max values for certain limiter types.
+ *
+ * Usage:
+ * This function is used for interpolating container values at faces in a finite volume method,
+ * ensuring that the interpolation adheres to the constraints imposed by the limiter.
```
Thanks. Removed the documentation at most places.

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   * @brief Functor for applying simplified slope limiting.
+   *
+   * This function applies the limiter by invoking the `limit` method with the provided parameters.
+   * It acts as a functor, enabling objects of the derived `Limiter` class to be used as if they
```
Updated. This extra context was removed

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   * RealVectorValue dCD(1.0, 0.0, 0.0);
+   * FaceInfo * fi = ... // Assume this is properly initialized
+   * Real result = (*myLimiter)(phi_upwind, phi_downwind, &grad_upwind, dCD);
+   * @endcode
```
Removed

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   * @brief Functor for applying general slope limiter.
+   *
+   * This function applies the slope limiter by invoking the `limit` method with the provided
+   * parameters. It acts as a functor, enabling objects of the `Limiter` class to be used as if they
```
Removed

tanoret commented on framework/include/limiters/Limiter.h:
```diff
@@ -61,23 +63,268 @@ class Limiter
+   * bool is_upwind = true;
+   * Real result = (*myLimiter)(phi_upwind, phi_downwind, &grad_upwind, &grad_downwind, dCD,
+   * max_value, min_value, fi, is_upwind);
+   * @endcode
```
Removed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2548241/), step Clang format on 5df2ba5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7377410027ba9dbf48902235a0437fdf96485e29`

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+ * @param phi_upwind The scalar value at the upwind location.
+ * @param phi_downwind The scalar value at the downwind location.
+ * @param grad_phi_upwind Pointer to the gradient vector at the upwind location.
+ * @param grad_phi_face Pointer to the gradient vector at the face location.
```
Thanks. Changed

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+ * element is upwind.
+ * 4. Computes the delta value at the face by taking the dot product of the upwind gradient vector
+ * with the difference between the face and cell centroids.
+ * 5. Returns the interpolated value by adding the product of the limiter \( \beta \) and the delta
```
Removed the description of what each function is doing

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+ * Usage:
+ * This function is typically used in the finite volume methods for min-max computations over
+ * stencils (neighborhoods). It helps compute the limiting for limited second order upwind at the
+ * faces.
```
Removed the description of what the function is doing

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+ * Usage:
+ * This function is typically used in the finite volume methods for min-max computations over
+ * stencils (neighborhoods). It helps compute the limiting for limited second order upwind at the
+ * faces.
```
Remove the description of the workflow of the function

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+ *
+ * Usage:
+ * This function is used for interpolating values at faces in a finite volume method, ensuring that
+ * the interpolation adheres to the constraints imposed by the limiter.
```
Removed the description of the workflow of the function

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
   typedef typename FunctorBase<T>::GradientType GradientType;
   static const GradientType zero(0);
 
+  // Assert that the face information is non-null
```
Removed this comment and similar ones below

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+                                 std::numeric_limits<Real>::min(),
+                                 *face.fi,
+                                 face.elem_is_upwind);
+  }
```
Corrected

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -586,12 +923,46 @@ interpCoeffsAndAdvected(const FunctorBase<T> & functor, const FaceArg & face, co
 }
 
+/**
+ * @brief Interpolates values using a specified functor and face argument.
```
Removed all briefs from all functions

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -538,45 +558,362 @@ interpolate(const Limiter & limiter,
+    if (face.limiter_type == LimiterType::Venkatakrishnan || face.limiter_type == LimiterType::SOU)
+    {
+      std::tie(max_value, min_value) = computeMinMaxValue(functor, face, *time_arg);
+    }
```
Corrected

tanoret commented on framework/include/utils/MathFVUtils.h:
```diff
@@ -586,12 +923,46 @@ interpCoeffsAndAdvected(const FunctorBase<T> & functor, const FaceArg & face, co
+ *
+ * Usage:
+ * This function is used for interpolating values at faces in a finite volume method, ensuring that
+ * the interpolation adheres to the constraints imposed by the limiter.
```
Removed all workflow summaries from the functions in this class

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2548394/), step Clang format on c6913a1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7377410027ba9dbf48902235a0437fdf96485e29`

grmnptr committed:
Disable recover due to comparison on final timestep Refs #28891

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2553316/) on c5dfb2c : invalidated by @grmnptr

grmnptr commented:
I dropped the submodule upodate and fixed some tests.

grmnptr commented:
I also did the technical review on this PR yesterday. As part of that, I ran the tests manually and noticed a few things:
- When I try some of the limiters with `Steady` and `previous_nl_solution_required = true` on the dispersion tests it actually converges somehow but to the wrong results. I suggest throwing errors if somebody wants to use the limiters with Steady executioners.
- The venkatakrishnan limiter doesn't seem to converge for the dispersion test on the rectangular grid (levels off at 1e-4ish), is this expected? This is the only test that has this limiter.
- None of the limiters converge for the dispersion test on the triangular mesh. They level off at around 1e-2 - 2e-2. When I refine the mesh (so use 150k elements in 2D instead of a few 10s) the behavior slightly improves, but still levels off at around 5e-4 - 1e-3. I tested the lid driven cavity example too with a triangular mesh and I found no convergence issues there (both SIMPLE and Transient converged to very low residuals even with higher Re numbers). I suppose this can be due to the discontinuity in the field, but would be nice to talk about this.

grmnptr commented:
I remembered wrong, considering that I did not run bigger, more application-oriented problems, my review is not a technical review. Just observations based on the test cases.

grmnptr committed:
Transition towards less limiting of SOU and move tests to framework Refs #28891

grmnptr committed:
Add error messages for steady state executioners when using second order limiting schemes Refs #28891

grmnptr commented:
Okay, I moved the limiter tests to the framework and added error messages for cases when the user selects steady executioners. I removed the triangular tests too. What is left at this point:
-  Regold the test files for SOU and the dispersion test and make sure that the gold files show up only once in the history

We may see issues additional issues with the new additions. I'll keep monitoring this PR.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2555238/), step Clang format on dba77af wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 37b999b88168e63b60d24edc34afeda04ee57b8a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2559905/), step Clang format on 04e3731 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8a5faa96a63686f29a61c4892f0f0683024be5ea`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 8cf736f690459f9829bbda58da5f4a76baa548c5
+Subproject commit 98c8751a0d2907d235d8e1e9ae6c54a867ab9380
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

tanoret committed:
regolding dispersion tests Refs #28891

grmnptr committed:
Modify linked design file in dispersion test.

grmnptr commented:
Okay, regold history is clean, we will need this in:
https://github.com/idaholab/moose/pull/29146
And then we can roll.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2568832/), step Clang format on c580b5b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28892/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format efd9ea4e203a5628c080244d6e57becc3ef322b2`

tanoret committed:
regolding for max_treads=1 in segreated lid-driven tests Refs #28891

grmnptr reviewed (approved):
Looks good to me, thanks for addressing all the comments on this PR! @GiudGiud if this is good to go from your side too we could merge this.

GiudGiud commented on framework/include/limiters/SOULimiter.h:
```diff
@@ -47,18 +47,25 @@ class SOULimiter : public Limiter<T>
+          const VectorValue<T> * /* grad_phi_downwind */,
           const RealVectorValue & dCD,
           const Real & max_value,
           const Real & min_value,
```
I think that's the only thing I'd try to change back and It does not need to be done now. 
Approved

tanoret replied:
Thanks. Note that SOU does not depend on the downwind gradients. The current implementation is correct

GiudGiud replied:
yes it's correct. It's just T vs Real that maybe we can do something about. Nothing that affects results

tanoret replied:
You mean for `max_value` and `min_value`? If so, it does make sense that they are real numbers. Limiting with the maximum value as a vector or ADReal does not really make sense imo

GiudGiud reviewed (changes requested):
actually this last commit should be squashed in a previous one
see
https://github.com/idaholab/moose/pull/28892/commits/dc5282514b9d8f5a2678a063c5131276b6a14669
https://github.com/idaholab/moose/pull/28892/commits/15691c99b6f9a23b5a9c19ee54f3c4b3452a69f4

you re checking in 8 gold files twice

tanoret commented:
> actually this last commit should be squashed in a previous one see [dc52825](https://github.com/idaholab/moose/commit/dc5282514b9d8f5a2678a063c5131276b6a14669) [15691c9](https://github.com/idaholab/moose/commit/15691c99b6f9a23b5a9c19ee54f3c4b3452a69f4)
> 
> you re checking in 8 gold files twice

I will delete the gold files from https://github.com/idaholab/moose/commit/15691c99b6f9a23b5a9c19ee54f3c4b3452a69f4 Finding a issues with squashing and don't have time to evaluate non-regression in the flow test bed

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 8cf736f690459f9829bbda58da5f4a76baa548c5
+Subproject commit fb009ce0b606195537fb3bdb03e338c5c49f39b0
```
**Caution!** This contains a submodule update

GiudGiud reviewed (approved):
test failures (FENIX build failure) due to old recipes (pyne issue) and unrelated
